day_5_supply_stacks/stack_parser.rb

```diff
@@ -0,0 +1,78 @@
+require 'minitest/autorun'
+
+class StackParser
+  def initialize(stack_input)
+    @stack_input = stack_input
+    @stack_rows = stack_rows
+  end
+
+  def stack_columns
+    raw_stack_columns = transpose_stack_rows
+    remove_empty_crates_from(raw_stack_columns)
+  end
+
+  private
+
+  attr_reader :stack_input, :stack_row
+
+  def transpose_stack_rows
+    rows = stack_rows.size
+    columns = stack_rows.max_by(&:size).size
+
+    raw_stack_columns = Array.new(columns) { Array.new(rows) }
+
+    stack_rows.each_with_index do |row, i|
+      row.each_with_index do |crate, j|
+        raw_stack_columns[j][i] = crate
+      end
+    end
+
+    raw_stack_columns
+  end
+
+  def remove_empty_crates_from(raw_stack_columns)
+    raw_stack_columns.map do |column|
+      column.delete_if { |crate| crate.nil? || crate.empty? }
+    end
+  end
+
+  def stack_rows
+    stack_input.map do |row|
+      row.each_char.each_slice(4).map do |crate|
+        crate.join('')
+             .gsub("[", "")
+             .gsub("]","")
+             .strip
+      end
+    end
+  end
+end
+
+class TestStackParser < MiniTest::Test
+  def test_intial_stack_setup
+    stack_input2 =[
+      "[N]     [Q]         [N]\n",
+      "[R]     [F] [Q]     [G] [M]\n",
+      "[J]     [Z] [T]     [R] [H] [J]\n",
+      "[T] [H] [G] [R]     [B] [N] [T]\n",
+      "[Z] [J] [J] [G] [F] [Z] [S] [M]\n",
+      "[B] [N] [N] [N] [Q] [W] [L] [Q] [S]\n",
+      "[D] [S] [R] [V] [T] [C] [C] [N] [G]\n",
+      "[F] [R] [C] [F] [L] [Q] [F] [D] [P]\n",
+    ]
```

Comment:
You can use a 'heredoc'. From our friend Jézus (not that one, but Jesus Castillo ;)):

https://www.rubyguides.com/2018/11/ruby-heredoc/

so, the above is could be written as 

```ruby
tack_input2 = <<MATRIX
[N]     [Q]         [N]
[R]     [F] [Q]     [G] [M]
[J]     [Z] [T]     [R] [H] [J]
[T] [H] [G] [R]     [B] [N] [T]
[Z] [J] [J] [G] [F] [Z] [S] [M]
[B] [N] [N] [N] [Q] [W] [L] [Q] [S]
[D] [S] [R] [V] [T] [C] [C] [N] [G]
[F] [R] [C] [F] [L] [Q] [F] [D] [P]
MATRIX
```

But that's not equivalent! "Mine is an `Array` and a heredoc is a `String`!" you cry!

Good observation, but we can do this:

```ruby
tack_input2 = <<MATRIX.split("\n")
[N]     [Q]         [N]
[R]     [F] [Q]     [G] [M]
[J]     [Z] [T]     [R] [H] [J]
[T] [H] [G] [R]     [B] [N] [T]
[Z] [J] [J] [G] [F] [Z] [S] [M]
[B] [N] [N] [N] [Q] [W] [L] [Q] [S]
[D] [S] [R] [V] [T] [C] [C] [N] [G]
[F] [R] [C] [F] [L] [Q] [F] [D] [P]
MATRIX
```

It's a bit weird syntax, but works!

It's a good idea to keep heredocs in your toolbox. it's almost always preferable for multi-line strings.

Reply:
So yeah, make sure to read https://www.rubyguides.com/2018/11/ruby-heredoc/ (e.g. the 'squiggly heredoc' is very useful in many cases)

Reply:
That is a good point -> I am a fan of heredoc. In this case, I argue for not using it, as in the tests, I want to see the input without scratching my head/thinking about it.